docs/design-documents/hal/0001-i2c-overhaul.md

```diff
@@ -0,0 +1,484 @@
+
+## Motivation
+
+Many of the current HAL API interfaces, including the I2C API, do not provide a clear, concrete specification to develop implementations against. As a result, the existing implementations lack consistency and do not have uniform behaviour across all partner boards.
+
+The inconsistencies in the current implementations due to poorly defined specifications are listed below, these will be fixed as part of this overhaul.
+
+As well as fixing existing inconsistencies there are a few unimplemented features that have been requested from the API, or unused features that have been overlooked, these will be added and removed respectively.
+
+Additionally, since the original implementations were added for the API a large number of additional boards have been supported, the overhaul to the API allows partners the opportunity to consolidate the implementations for a family of boards into a single implementation, which will reduce the development overhead in future if any changes are made down the line.
+
+### Inconsistencies
+
+- The implementation for `i2c_reset` varies across all platforms.
+
+  The purpose of this function is to add a method of resetting the i2c bus. Most platform implementations do not implement this function correctly, either the implementation acts as a NOP and does nothing, or it calls `i2c_stop` which already has a functional representation in the API and is designed to perform a different action.
+
+- The implementation of the `i2c_start` and `i2c_stop` functions are not well defined.
+
+  The return value of the function is not defined in the header file, the value has been left open to interpretation by the partner implementations, for example, some implementations return 0 on all conditions. Some implementations have added timeouts to the functionality of the function and return 1 on timeouts. This needs to be defined correctly to be uniform across all implementations.
+
+- The `i2c_slave_address` function has a parameter called `mask`, this parameter is listed as unused but some implementations use it to mask the `address` parameter.
+
+  This parameter should be removed as it is not required by the API. It can be assumed that the address will be masked to exclude the read/write bit from the address, the behaviour of this does not need to change.
+
+- The `i2c_slave_address` function has a parameter called `idx`, this parameter is listed as unused but is utilised in some API implementations.
+
+  This parameter is not required by the API, it should be removed to keep behaviour uniform across all platforms.
+
+- The implementation of the `i2c_byte_write` differs between platforms.
+
+  The API documentation indicates that the return value should be 0 on success, 1 on NAK, 2 on timeout. Not all platforms return a value or handle a timeout. Some platforms return a completely different set of status, the Atmel implementation returns an `I2C_ERROR ` which does not map to the documented return value.
```

Comment:
The return value mixup also show up on the I2C class, as byte write and multiple byte write return different status values on ack/nack.

Reply:
Yeah, the issue exists there too, however, the driver API is outside the scope of the HAL API overhaul.